File: ios/AIRTwitter/Functions/SendDirectMessageFunction.m
/**
 * Copyright 2015 Marcel Piestansky (http://marpies.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "SendDirectMessageFunction.h"
#import "MPFREObjectUtils.h"
#import "MPStringUtils.h"
#import "AIRTwitterEvent.h"
#import "StatusUtils.h"
#import "AIRTwitter.h"
#import "DirectMessageUtils.h"

FREObject tw_sendDirectMessage( FREContext context, void* functionData, uint32_t argc, FREObject* argv ) {
    NSString* text = [MPFREObjectUtils getNSString:argv[0]];
    double userIDDouble = [MPFREObjectUtils getDouble:argv[1]];
    NSString* userID = (userIDDouble >= 0) ? [NSString stringWithFormat:@"%.f", userIDDouble] : nil;
    NSString* screenName = (argv[2] == nil) ? nil : [MPFREObjectUtils getNSString:argv[2]];
    int callbackID = [MPFREObjectUtils getInt:argv[3]];
    
    [[AIRTwitter api] postDirectMessage:text
                          forScreenName:screenName
                               orUserID:(screenName ? nil : userID)
                           successBlock:^(NSDictionary *message) {
//                               [AIR log:[NSString stringWithFormat:@"Success sending DM %@", message]];
                               NSMutableDictionary* dmJSON = [DirectMessageUtils getJSON:message];
                               dmJSON[@"callbackID"] = @(callbackID);
                               dmJSON[@"success"] = @(true);
                               NSString* resultJSON = [MPStringUtils getJSONString:dmJSON];
                               if( resultJSON ) {
                                   [AIRTwitter dispatchEvent:DIRECT_MESSAGE_QUERY_SUCCESS withMessage:resultJSON];
                               } else {
                                   [AIRTwitter dispatchEvent:DIRECT_MESSAGE_QUERY_SUCCESS withMessage:[MPStringUtils getEventErrorJSONString:callbackID errorMessage:@"Succesfully sent direct message but could not parse returned message data."]];
                               }
                           }
                             errorBlock:^(NSError *error) {
                                 [AIRTwitter log:[NSString stringWithFormat:@"Error sending DM %@", error.localizedDescription]];
                                 [AIRTwitter dispatchEvent:DIRECT_MESSAGE_QUERY_ERROR withMessage:[MPStringUtils getEventErrorJSONString:callbackID errorMessage:error.localizedDescription]];
                             }];
    
    return nil;
}

File: android/src/com/marpies/ane/twitter/functions/SendDirectMessageFunction.java
/**
 * Copyright 2015-2016 Marcel Piestansky (http://marpies.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.marpies.ane.twitter.functions;

import com.adobe.fre.FREContext;
import com.adobe.fre.FREObject;
import com.marpies.ane.twitter.data.AIRTwitterEvent;
import com.marpies.ane.twitter.data.TwitterAPI;
import com.marpies.ane.twitter.utils.*;
import twitter4j.*;

public class SendDirectMessageFunction extends BaseFunction {

	@Override
	public FREObject call( FREContext context, FREObject[] args ) {
		super.call( context, args );

		String message = FREObjectUtils.getString( args[0] );
        String userID = (args[1] == null) ? null : FREObjectUtils.getString( args[1] );
		String screenName = (args[2] == null) ? null : FREObjectUtils.getString( args[2] );
		mCallbackID = FREObjectUtils.getInt( args[3] );

		AsyncTwitter twitter = TwitterAPI.getAsyncInstance( TwitterAPI.getAccessToken() );
		twitter.addListener( this );
		if( screenName != null ) {
			twitter.sendDirectMessage( screenName, message );
		} else {
			twitter.sendDirectMessage( Long.valueOf( userID ), message );
		}

		return null;
	}

	@Override
	public void sentDirectMessage( DirectMessage message ) {
		AIR.log( "Success sending DM '" + message.getText() + "'" );
		try {
			JSONObject dmJSON = DirectMessageUtils.getJSON( message );
			dmJSON.put( "listenerID", mCallbackID );
			dmJSON.put( "success", "true" );
			AIR.dispatchEvent( AIRTwitterEvent.DIRECT_MESSAGE_QUERY_SUCCESS, dmJSON.toString() );
		} catch( JSONException e ) {
			e.printStackTrace();
			AIR.dispatchEvent( AIRTwitterEvent.DIRECT_MESSAGE_QUERY_SUCCESS,
					StringUtils.getEventErrorJSON( mCallbackID, e.getMessage() )
			);
		}
	}

	@Override
	public void onException( TwitterException te, TwitterMethod method ) {
		if( method == TwitterMethod.SEND_DIRECT_MESSAGE ) {
			AIR.log( "Error trying to send DM: " + te.getMessage() );
			AIR.dispatchEvent( AIRTwitterEvent.DIRECT_MESSAGE_QUERY_ERROR,
					StringUtils.getEventErrorJSON( mCallbackID, te.getMessage() )
			);
		}
	}
}
